File: src/Gweb.Shared/Correlation/CorrelationContext.cs
namespace Gweb.Shared.Correlation;

public sealed record CorrelationContext(string CorrelationId, string Handler, string? ApplicationId = null);

/// <summary>
/// One execution environment handles one invocation at a time (AWS Lambda's execution
/// model), so a static AsyncLocal is safe: it never leaks context between concurrent
/// requests the way a naive static mutable field would on a real multi-request server.
/// </summary>
public static class CorrelationScope
{
    private static readonly AsyncLocal<CorrelationContext?> Current = new();

    public static T Run<T>(CorrelationContext context, Func<T> fn)
    {
        var previous = Current.Value;
        Current.Value = context;
        try
        {
            return fn();
        }
        finally
        {
            Current.Value = previous;
        }
    }

    public static CorrelationContext? GetCurrent() => Current.Value;
}

File: shared/correlation/correlationContext.ts
import { AsyncLocalStorage } from "node:async_hooks";

export interface CorrelationContext {
  correlationId: string;
  applicationId?: string;
  handler: string;
}

// One execution environment handles one invocation at a time (AWS Lambda's execution
// model), so a module-level AsyncLocalStorage is safe: it never leaks context between
// concurrent requests the way a naive global variable would.
const storage = new AsyncLocalStorage<CorrelationContext>();

export function runWithCorrelationContext<T>(context: CorrelationContext, fn: () => T): T {
  return storage.run(context, fn);
}

export function getCorrelationContext(): CorrelationContext | undefined {
  return storage.getStore();
}
